2.9.0 release blog (#1001)
kedacore/keda-docs: content/blog/2022-12-12-keda-2.9.0-release.md

+++
title = "Announcing KEDA v2.9 🎉"
date = 2022-12-12

> author = "Jeff Hollan (KEDA), Tom Kerkhove (KEDA)"

Reviewer: ```suggestion
author = "Jeff Hollan (Snowflake), Tom Kerkhove (Microsoft) and Zbynek Roubalik (Red Hat)"
```

- **Azure Service Bus Scaler:** Support for Shared Access Signature (SAS) tokens for authentication and use of regex in `queueName` / `subscriptionName`
- **ElasticSearch Scaler:** Support for Elastic Cloud
- **GCP Storage Scaler:** Support for blob prefix & delimiters
- **Metrics API Scaler:** Support for unsafeSsl parameter to skip certificate validation when connecting over HTTPS
- **NATS Jetstream Scaler:** Support for clustered scenarios
- **NATS Scalers:** Support HTTPS protocol
- **Prometheus Scaler:** Introduce skipping of certificate check for unsigned certs
- **Pulsar Scaler:** Add support for basic and bearer token authentication along with support for partitioned topics

Details on the underlying fixes can be found in the [changelog](https://github.com/kedacore/keda/blob/main/CHANGELOG.md#v290).

## Production-grade operating experience for autoscaling applications

There are a number of improvements into the ways KEDA provides operational information in Prometheus and how you can consume it.

> For starters, our operator now exposes Prometheus metrics which can be scraped on port `8080` at `/metrics`. It provides the same set of metrics as the metric server (although renamed) offers today, in addition to a few new metrics and the introduction of a `ScalerName` label.

Reviewer: ```suggestion
For starters, **our operator now exposes Prometheus metrics** which can be scraped on port `8080` at `/metrics`. It provides the same set of metrics as the metric server (although renamed) offers today, in addition to a few new metrics and the introduction of a `ScalerName` label.
```

The metrics are moving given we are changing how KEDA is integrating with external systems so that we can improve the performance and reduce the load on those external systems.

> As part of this, we are deprecating the KEDA-specific metrics in our metric server and recommend end-users to use the metrics in the operator. Join our [conversation there](https://github.com/kedacore/keda/discussions/3973).

Reviewer: ```suggestion
As part of this, **we are deprecating the KEDA-specific metrics in our metric server** and recommend end-users to use the metrics in the operator. Join our [conversation there](https://github.com/kedacore/keda/discussions/3973).
```

> Next, we have introduced `keda_resource_totals`and `keda_trigger_totals` that provide insights on your autoscaling landscape and the various autoscaling triggers that are being used.

Reviewer: ```suggestion
Next, we have **introduced `keda_resource_totals` and `keda_trigger_totals` as metrics** that provide insights into your autoscaling landscape and the various autoscaling triggers that are being used.
```

> 
> Metrics are great - But they have to be consumed. That's why we introduced a new Grafana dashboard allows you to easily visualize the KEDA metrics in Grafana to see how your applications are scaling:

Reviewer: ```suggestion

Metrics are great - But they have to be consumed. That's why we **introduced a new Grafana dashboard allows end-users to easily visualize the KEDA metrics in Grafana** to see how your applications are scaling:
```

![Grafana Dashboard](/img/blog/v2.9-release/grafana-dashboard.png)

> Lastly, end-users using `ScaledJobs` can now see what the minimum replica count is when inspecting resources through the Kubernetes API.

Reviewer: ```suggestion
Lastly, **end-users using `ScaledJobs` can now see what the minimum replica count is when inspecting resources** through the Kubernetes API.
```

Learn more about our available Prometheus metrics and our Grafana dashboard in our [docs](https://keda.sh/docs/operate/prometheus/).

## Introduce capability to cache metric values for a scaler during the polling interval (experimental feature)

> TODO: @zroubalik

Reviewer: @zroubalik

> ## Introduce capability to cache metric values for a scaler during the polling interval (experimental feature)
> 
> TODO: @zroubalik

Reviewer: ```suggestion
## Change in the architecture and metrics caching

In this release we focused on performance improvements. Let's talk about this changes: KEDA has two main components - the Operator and the Metrics Server. The Operator monitors ScaledObjects and ScaledJobs resources, and checks activity on the external service. The Metrics Server handles requests and queries about metric values used for scaling from Kubernetes' Horizontal Pod Autoscaler (HPA) by checking the external service.

In previous versions of KEDA, each component needed to handle its own connection to the external service. In version 2.9, we have changed this and moved the connection logic to the Operator component. We are now using gRPC for the connection between the Metrics Server and Operator. This has several benefits, including reduced load on the external service, as there is only one connection from the Operator.

We have also introduced an experimental feature called metrics value caching. This caches the values in the Operator and can be reused in future. It can even further reduce the load on the external service. By default, Kubernetes' HPA asks for a metric value every few seconds (as defined b `--horizontal-pod-autoscaler-sync-period` flag set by Kuberentes admins, which is usually set to 15 seconds). This can create unnecessary load on the external service. Caching metric values per trigger in a ScaledObject can help to overcome this problem. This is an optional feature, and can be enabled by setting the `useCachedMetrics` flag, then metric values are cached during polling interval as defined by `pollingInterval` flag and request comming from HPA to KEDA Metrics Server are served from this cache.

We are excited about the potential uses for this feature in the future, including using AI and machine learning models to affect metric values. Stay tuned for more updates and improvements in future releases of KEDA.
```

## New ways to authenticate

In this release we have added some additional configuration options for triggers:

Azure Key Vault authentication provider now allows end-users to use pod identities for authentication for improved simplicity & security:

> ```

Reviewer: ```suggestion
```diff
```

+++
title = "Announcing KEDA v2.9 🎉"
date = 2022-12-12
author = "Jeff Hollan (Snowflake), Tom Kerkhove (Microsoft) and Zbynek Roubalik (Red Hat)"
aliases = [
"/blog/releases/2.9.0"
]
+++

We recently completed our most recent release: 2.9.0 🎉!

Here are some highlights:

- Newly published Deprecations and Breaking Change policy ([docs](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md))
- Introduce new CouchDB, Etcd & Loki scalers
- Introduce off-the-shelf Grafana dashboard for application autoscaling
- Introduce improved operational metrics in Prometheus

> - Introduce capability to cache metric values for a scaler during the polling interval (experimental feature)

Reviewer: ```suggestion
- Introduce capability to cache metric values for a scaler during the polling interval (experimental feature)
- Performance improvements and architecture changes on how metrics are exposed to Kuberentes  ([issue]https://github.com/kedacore/keda/issues/2282)
```

As part of this, **we are deprecating the KEDA-specific metrics in our metric server** and recommend end-users to use the metrics in the operator. Join our [conversation there](https://github.com/kedacore/keda/discussions/3973).

Next, we have **introduced `keda_resource_totals` and `keda_trigger_totals` as metrics** that provide insights into your autoscaling landscape and the various autoscaling triggers that are being used.

Metrics are great - But they have to be consumed. That's why we **introduced a new Grafana dashboard allows end-users to easily visualize the KEDA metrics in Grafana** to see how your applications are scaling:

![Grafana Dashboard](/img/blog/v2.9-release/grafana-dashboard.png)

Lastly, **end-users using `ScaledJobs` can now see what the minimum replica count is when inspecting resources** through the Kubernetes API.

Learn more about our available Prometheus metrics and our Grafana dashboard in our [docs](https://keda.sh/docs/operate/prometheus/).

## Change in the architecture and metrics caching

> In this release we focused on performance improvements. Let's talk about this changes: KEDA has two main components - the Operator and the Metrics Server. The Operator monitors ScaledObjects and ScaledJobs resources, and checks activity on the external service. The Metrics Server handles requests and queries about metric values used for scaling from Kubernetes' Horizontal Pod Autoscaler (HPA) by checking the external service.

Reviewer: ```suggestion
In this release we focused on performance improvements. Let's talk about these changes: KEDA has two main components - the Operator and the Metrics Server. The Operator monitors ScaledObjects and ScaledJobs resources, and checks activity on the external service. The Metrics Server handles requests and queries about metric values used for scaling from Kubernetes' Horizontal Pod Autoscaler (HPA) by checking the external service.
```

We are excited about the potential uses for this feature in the future, including using AI and machine learning models to affect metric values. Stay tuned for more updates and improvements in future releases of KEDA.

## New ways to authenticate

In this release we have added some additional configuration options for triggers:

Azure Key Vault authentication provider now allows end-users to use pod identities for authentication for improved simplicity & security:

```diff
azureKeyVault:                                          # Optional.
  vaultUri: https://keda.vault.azure.net

> + podIdentity:                                          # Optional.
> +   provider: azure-workload                    # Required.

Reviewer: ```suggestion
+ podIdentity:
+   provider: azure-workload
```
